keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,390 @@
+is causing several issues today.
+
+* System components like Cluster Autoscaler and Scheduler have to combine several Node Conditions to evaluate a node’s
+actual readiness. It is not straightforward to introduce new conditions into the policy.
+* External systems (such as networking by the cloud provider) must often be configured before pods should be scheduled on a node.
+* Cluster Autoscaler incorrectly assumes a GPU node to be “ready” prior to device plugins and/or device drivers
+being installed which leads to spurious node additions and deletions thereby leading to custom logic in the autoscaler.
+* Logs & metrics are lost if logging & monitoring agents are not online which can lead to sub-optimal (often unexpected)
+user behavior.
+* A node with offline kube-proxy can result in service outage in the event of an application update (via deployments for example).
+* It is difficult for Node Lifecycle Controller and/or node repair systems to detect and remedy issues with extended system
+components of the node since they lack a set of generic & easy to identify signals from those individual components on the node.
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems,
+it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal introduces an approach for addressing the issue of node health extensibility using
+node readiness gates.
+
+### Goals
+
+* introduce a mechanism that allows for customizable/extensible node readiness checks.
+
+### Non-Goals
+
+* improving time-to-readiness of nodes
+
+## Proposal
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+* Be backwards compatible
+* Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+* Ideally, provide a single signal (not a policy) to cluster level components.
+* Ideally, avoid introducing new API primitives.
+* Support self-bootstrapping - scheduling system components (including the kubelet) even if the node is not logically ready.
+
+## Design Details
+
+### Determining Readiness via Pod Selectors
+
+In this design, the indication of node readiness can be expressed by the readiness/health of specific pods running on that node. This approach
+would leverage existing mechnanisms and well-known concepts to determine if a node is ready to receive pods.
+
+A readiness taint will be used to track the overall readiness of a node. The kubelet, node lifecycle controller, other external controllers will
+manage that taint.
+
+The Node object will be extended to include a list of readiness gates like so:
+
+```go
+type NodeSpec struct {
+  ...
+  // Defines the list of readiness gates for a node
+  // +optional
+  ReadinessGates []NodeReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,7,opt,name=readinessGates"`
```

Comment:
If we ignore the implementation details for a moment, this API is not necessary.
The set of pods to wait for is already clearly specified by users: it is the set of pods with a `system-cluster-critical` or `system-node-critical` `priorityClass`.
We shouldn't need any additional input.

I feel that this `ReadinessGates` API would just be used to replicate the information already specified in `priorityClass`: if a node has a critical `priorityClass` it should be matched by a `ReadinessGates`, and vice-versa I can't think of a case where a pod matched by a `ReadinessGates` shouldn't have a critical `priorityClass`.

It seems that introducing `ReadinessGates` is driven by the desire to implement this in `kubelet`. However, if the implementation involved the scheduler to determine the set of pods with a critical `priorityClass` to be scheduled on the node, that new API wouldn't be necessary.

Reply:
I discussed this with @vishh. I was making the assumption that we'd have to deal with only one scheduler, because I was considering only daemonset pods. But if we want to deal with pods other than daemonsets, which are scheduled by different schedulers, then this becomes complicated because all schedulers would need to support this logic.